librz/bin/pdb/pdb.c

```diff
@@ -1,1433 +1,579 @@
- * 
- * \param pdb PDB information
- * \param mode printing mode
- */
-static void print_types(const RzPdb *pdb, PJ *pj, const int mode) {
-	RzList *plist = pdb->pdb_streams;
-	STpiStream *tpi_stream = rz_list_get_n(plist, ePDB_STREAM_TPI);
 
-	if (!tpi_stream) {
-		eprintf("There is no tpi stream in current pdb\n");
-		return;
+	ut32 stream_directory_len = block_num * pdb->super_block->block_size;
+	ut8 *stream_directory = (ut8 *)malloc(stream_directory_len);
+	if (!stream_directory) {
+		RZ_FREE(block_map);
+		goto error_memory;
 	}
-	switch (mode) {
-	case 'd': print_types_regular(pdb, tpi_stream->types); return;
-	case 'j': print_types_json(pdb, pj, tpi_stream->types); return;
-	case 'r': print_types_format(pdb, tpi_stream->types); return;
+	for (size_t i = 0; i < block_num; i++) {
+		rz_buf_seek(pdb->buf, (long long)block_map[i] * pdb->super_block->block_size, RZ_BUF_SET);
+		rz_buf_read(pdb->buf, stream_directory + i * pdb->super_block->block_size, pdb->super_block->block_size);
 	}
-}
-
-static void print_gvars(RzPdb *pdb, ut64 img_base, PJ *pj, int format) {
-	SStreamParseFunc *omap = 0, *sctns = 0, *sctns_orig = 0, *gsym = 0, *tmp = 0;
-	SIMAGE_SECTION_HEADER *sctn_header = 0;
-	SGDATAStream *gsym_data_stream = 0;
-	SPEStream *pe_stream = 0;
-	SGlobal *gdata = 0;
-	RzListIter *it = 0;
-	RzList *l = 0;
-	char *name;
-
-	l = pdb->pdb_streams2;
-	rz_list_foreach (l, it, tmp) {
-		switch (tmp->type) {
-		case ePDB_STREAM_SECT__HDR_ORIG:
-			sctns_orig = tmp;
-			break;
-		case ePDB_STREAM_SECT_HDR:
-			sctns = tmp;
-			break;
-		case ePDB_STREAM_OMAP_FROM_SRC:
-			omap = tmp;
-			break;
-		case ePDB_STREAM_GSYM:
-			gsym = tmp;
-			break;
-		default:
-			break;
-		}
-	}
-	if (!gsym) {
-		eprintf("There is no global symbols in current PDB.\n");
-		return;
+	RzBuffer *sd = rz_buf_new_with_bytes(stream_directory, stream_directory_len);
+	if (!sd) {
+		RZ_FREE(stream_directory);
+		RZ_FREE(block_map);
+		goto error_memory;
 	}
+	RZ_FREE(block_map);
 
-	if (format == 'j') {
-		pj_ka(pj, "gvars");
+	MsfStreamDirectory *msd = RZ_NEW0(MsfStreamDirectory);
+	if (!msd) {
+		goto error_memory;
 	}
-	gsym_data_stream = (SGDATAStream *)gsym->stream;
-	if ((omap != 0) && (sctns_orig != 0)) {
-		pe_stream = (SPEStream *)sctns_orig->stream;
-	} else {
-		if (sctns) {
-			pe_stream = (SPEStream *)sctns->stream;
-		}
+	if (!rz_buf_read_le32(sd, &msd->NumStreams)) {
+		RZ_FREE(msd);
+		goto error;
 	}
-	if (!pe_stream) {
-		return;
+	msd->StreamSizes = (ut32 *)malloc(msd->NumStreams * sizeof(ut32));
+	msd->sd = sd;
+	if (!msd->StreamSizes) {
+		RZ_FREE(msd);
+		goto error_memory;
+	}
+	ut32 total_blocks = 0;
+	for (size_t i = 0; i < msd->NumStreams; i++) {
+		ut32 stream_size;
+		if (!rz_buf_read_le32(sd, &stream_size)) {
+			RZ_FREE(msd);
+			goto error;
+		}
+		msd->StreamSizes[i] = stream_size;
+		ut32 blocks = count_blocks(stream_size, pdb->super_block->block_size);
+		total_blocks += blocks;
+	}
+	//				NumStreams 						StreamsSizes 				StreamsBlockMap
```

Comment:
What does this comment mean?

Reply:
Indicate the meaning of each sizeof(ut32)